### PR TITLE
Add pre-commit hook to automate local testing for commits

### PR DESCRIPTION
**Pull request by Lionex**
I started using a `pre-commit` hook on my own repositories and it's made my life much easier by automating the process of running tests; because it stashes all of the unstaged changes, it prevents bugs from cropping up because I've forgotten to add some code to a respective commit that my changes depend on.  It also means I don't have to remember every command I need to use for testing, so if I've forgotten to test with a feature flag this script will catch that for me.

I added instructions for setting up the `pre-commit` hook to CONTRIBUTING.md, but I'm not sure I'm totally satisfied with them.

As for the content of the script, I considered adding `cargo fmt` but it would fail because some of the example and source files were longer than 100 lines, and it introduced a whole host of changes to the source code that I wasn't sure how to deal with.

Finally, I'm not exactly an expert with shell scripts, so if there are changes that can make the script better please suggest away!

### CODE REVIEW

**Comment by torkleyy** [comment]
I really like this idea! Already thought myself about this, will review in four hours.

**Comment by Lionex** [comment]
@torkleyy I guess we are on the same wavelength then!

I found a bug with exit codes from the tests failing to propagate, fixed it now; if you need me to squash the commits I can go ahead and do that.

I think there are also lots of other improvements to be made, so I'm open to feedback!

File: .pre-commit.sh [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+STASH_NAME="pre-commit-$(date +%s)"
+RED='\033[1;31m'
+GREEN='\033[1;32m'
+YELLOW='\033[1;33m'
+NC='\033[0m'
+
+if git stash save -u -k -q $STASH_NAME; then
+    echo "${YELLOW}Stashed changes as:${NC} ${STASH_NAME}\n\n"
+    stash=1
```
Comment (vitvakatu) [inline .pre-commit.sh]: I guess you need to set `stash` to 0 in the beginning.

Comment (torkleyy) [inline .pre-commit.sh]: @vitvakatu Why that? Without setting the `stash` flag the changes aren't applied afterwards.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -127,6 +127,28 @@ done the following things first:
...
 
+To streamline the process, the `.pre-commit.sh` script will run tests on each
+commit in a sanitary way for you; to have it run automatically on each commit
+simply create a symbolic link with
+
+```bash
+$ ln -s -f ../../.pre-commit.sh .git/hooks/pre-commit
+```
+
+This ensures that you can't commit your changes if tests fail.  If you need to
```
Comment (torkleyy) [inline CONTRIBUTING.md]: Two spaces

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -127,6 +127,28 @@ done the following things first:
...
+make a commit without running tests, then simply pass the `--no-verify` flag.
+We recommend the following aliases:
+
+```bash
```
Comment (torkleyy) [inline CONTRIBUTING.md]: Wouldn't necessarily recommend these in a `CONTRIBUTING.md`; users not familiar to `git` aliases won't know what they did and how to commit.

File: .pre-commit.sh [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+fi
+
```
Comment (torkleyy) [inline .pre-commit.sh]: Not sure if we should do / ask for a rebase?

File: .pre-commit.sh [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+cargo build &&
+# Build and test without profiler
+cargo test --all &&
+# Build and test with profiler
+cargo test --all --features profiler
+
+# Capture exit code from tests
+status=$?
+
+if [ "$stash" -eq 1 ]
```
Comment (torkleyy) [inline .pre-commit.sh]: Good idea to stash them!

File: .pre-commit.sh (hunk 1 above)
Comment (vitvakatu) [inline .pre-commit.sh]: @torkleyy what if I have `stash` flag in my environment? For example after running `export stash=1`?

Comment (torkleyy) [inline .pre-commit.sh]: See [The Open Group - Environment Variables](http://pubs.opengroup.org/onlinepubs/000095399/basedefs/xbd_chap08.html):

> Environment variable names [...] consist solely of uppercase letters

It's not a requirement, but by convention, environment variables are UPPERCASE_ONLY.

File: CONTRIBUTING.md (hunk 3 above)
Comment (Lionex) [inline CONTRIBUTING.md]: Should I just mention the `--no-verify` flag as an alternative?

File: .pre-commit.sh (hunk 1 above)
Comment (vitvakatu) [inline .pre-commit.sh]: I can be stupid enough or not experienced enough. Well, I will approve it, but still, imo it's not problem to init variable before use.

File: CONTRIBUTING.md (hunk 3 above)
Comment (torkleyy) [inline CONTRIBUTING.md]: Yes, I think that is really useful.

File: .pre-commit.sh (hunk 1 above)
Comment (Lionex) [inline .pre-commit.sh]: I tried setting `stash` to 0 at the beginning and it seemed to change the behavior; now, if i test the script on a commit with no staged changes and a clean working directory, the script stashes the changes and then when it tries to un-stash it fails and results in unmerged paths, requiring me to `git reset HEAD -- && git checkout .` to get back to my pre-commit state.

Comment (torkleyy) [inline .pre-commit.sh]: @vitvakatu sorry I missed the "in the beginning" part of your comment! You can of course do that. Still, I never saw that because if you don't have an environment `stash`.

Comment (torkleyy) [inline .pre-commit.sh]: @Lionex that's strange.

Comment (Lionex) [inline .pre-commit.sh]: @torkleyy I did some more testing and it only appears if working directory is clean; I guess it's a rare edge case, should we worry about it?

Comment (Lionex) [inline .pre-commit.sh]: @torkleyy I figured it out, if there are no local changes to stash, `git stash save` exits without error and sets `stash` to 1; later the script attempts to unstash something I saved from another branch causing unmerged paths.  

That doesn't explain why it doesn't happen if my commit isn't empty though.

File: .pre-commit.sh (hunk 4 above)
Comment (Lionex) [inline .pre-commit.sh]: If people do make the symbolic link and switch to a branch in the project that doesn't include the script their `pre-commit` hook won't run.  I don't know if that prevents committing though and requires the `--no-verify` flag..

File: .pre-commit.sh (hunk 1 above)
Comment (torkleyy) [inline .pre-commit.sh]: If not setting `stash` to zero solves it, I'd just leave it out.

File: .pre-commit.sh (hunk 4 above)
Comment (torkleyy) [inline .pre-commit.sh]: @Lionex What does that have to do with rebasing? Right, the symbolic link is problematic in this case, what about `cp`ing it?

File: .pre-commit.sh (hunk 1 above)
Comment (Lionex) [inline .pre-commit.sh]: Actually it appears in both cases, I had just been testing it differently when I wrote the script the first time.

File: .pre-commit.sh (hunk 4 above)
Comment (Lionex) [inline .pre-commit.sh]: I guess I misunderstood the context for mentioning rebasing.

On my other repositories I'm using `cp` until the hook shows up in all of my branches.  That _does_ mean that updating it can be a pain though.

File: .pre-commit.sh (hunk 1 above)
Comment (torkleyy) [inline .pre-commit.sh]: What about using `git diff-index --quiet HEAD --` (not sure if that's correct) to determine if there are are untracked changes?

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -127,6 +127,36 @@ done the following things first:
...
+$ ln -s -f ../../.pre-commit.sh .git/hooks/pre-commit
+```
+
+or copy it to your `.git` folder with
+
+```bash
+$ cp .pre-commit.sh .git/hooks/pre-commit
+```
+
+Note: if you use `cp` you won't get upstream changes, but if you use `ln` and
```
Comment (torkleyy) [inline CONTRIBUTING.md]: Would adding it to `.gitignore` work?

File: .pre-commit.sh [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+# Build and test without profiler
+cargo test --all &&
+# Build and test with profiler
+cargo test --all --features profiler
+
+# Capture exit code from tests
+status=$?
+
+# Revert stash if changes were stashed to restor working directory files
+if [ "$stash" -eq 1 ]
```
Comment (torkleyy) [inline .pre-commit.sh]: stash is "" if it's an empty commit

File: .pre-commit.sh [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+STASH_NAME="pre-commit-$(date +%s)"
+RED='\033[1;31m'
+GREEN='\033[1;32m'
+YELLOW='\033[1;33m'
+NC='\033[0m'
+
+# Check to make sure commit isn't emtpy
+if git diff-index --quiet HEAD --; then
+    echo "${RED} Empty commit"
```
Comment (torkleyy) [inline .pre-commit.sh]: probably exit the pre-commit hook?

File: .pre-commit.sh [hunk 9]
```diff
@@ -0,0 +1,42 @@
+#!/usr/bin/env sh
+# pre-commit.sh
+
+STASH_NAME="pre-commit-$(date +%s)"
+RED='\033[1;31m'
+GREEN='\033[1;32m'
+YELLOW='\033[1;33m'
+NC='\033[0m'
+
```
Comment (torkleyy) [inline .pre-commit.sh]: Would be nice to have some informative message like "Testing commit"

File: .pre-commit.sh [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+# Revert stash if changes were stashed to restor working directory files
+if [ "$stash" -eq 1 ]
+then
+    if git stash pop -q; then
+        echo "\n\n${GREEN}Reverted stash command${NC}"
+    else
+        echo "\n\n${RED}Unable to revert stash command${NC}"
+    fi
+fi
+
```
Comment (torkleyy) [inline .pre-commit.sh]: Would be nice to tell the user it didn't quite work out:

```bash
if [ "$status" -eq "0" ]
then
    echo "Build failed; if you still want to commit use '--no-verify'"
fi
```

File: CONTRIBUTING.md (hunk 6 above)
Comment (Lionex) [inline CONTRIBUTING.md]: I don't think it would have any effect

**Comment by Lionex** [comment]
@torkleyy Thanks for the suggestions, incorporating them now!

File: .pre-commit.sh (hunk 10 above)
Comment (Lionex) [inline .pre-commit.sh]: They still see all of the test output but I'll add it right now, the suggestion to use `--no-verify` would be especially helpful

File: CONTRIBUTING.md (hunk 6 above)
Comment (torkleyy) [inline CONTRIBUTING.md]: Ah you mean upstream changes to the `pre-commit` hook, now it makes sense.

**Comment by Lionex** [comment]
@torkleyy Are contributors on windows not using something like `git bash` in the first place?

**Comment by torkleyy** [comment]
@Lionex if there is a bash shell executed on Windows, I'm perfectly fine with this PR. It should work and if it does not, we can still fix it later when someone notices an issue.

File: .pre-commit.sh [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+    echo "\tMake sure to add your changes with 'git add'"
+    exit 1
+else
+    # Stash all changes in the working directory so we test only commit files
+    if git stash save -u -k -q $STASH_NAME; then
+        echo "${YELLOW}Stashed changes as:${NC} ${STASH_NAME}\n\n"
+        stash=1
+    fi
+fi
+
```
Comment (torkleyy) [inline .pre-commit.sh]: Oh I forgot, what about a rebase here?

Comment (Lionex) [inline .pre-commit.sh]: Good catch!  
I found a [blog post](http://gmurphey.com/2013/02/02/ignoring-git-hooks-when-rebasing.html) with a solution, will test it when I get home.

**Comment by Lionex** [comment]
@torkleyy I tested on a rebase, the updated script doesn't interfere.

**Comment by torkleyy** [comment]
@Lionex :+1:

I thought a bit more about this and since the advantage of this is that we don't have travis fail anymore, it would probably be nice to check if a rebase is required and if so, prompt the user to do that

**Comment by Lionex** [comment]
@torkleyy What would require a rebase?  I think that may be more appropriate for a pre-push hook.

**Comment by Lionex** [comment]
I'm down to write a `pre-push` hook whenever I have time again.  Finals are coming up and I might be starting a new job, but if we can lay out exactly what situation calls for a `rebase` (perhaps in an issue?) I'll get to it as soon as possible and open a separate PR for it.

**Comment by torkleyy** [comment]
@Lionex thank you for efforts! The pre-push hook is definitely not a high priority. I think we can merge this PR as is.
